File: dP/PayItems.java
package dP;

import java.util.ArrayList;

interface PayMethod {
    void pay(int amountP);
}

class CreditCard implements PayMethod {

    private String name;
    private long cardNumber;
    private String cvv;
    private String dateOfExpiry;

    CreditCard(String nm, long cN, String cvvP, String dOE) {
        name = nm;
        cardNumber = cN;
        cvv = cvvP;
        dateOfExpiry = dOE;
    }

    @Override
    public void pay(int amountP) {
        System.out.println(amountP + " paid using credit/debit card.");
    }
}

class Paypal implements PayMethod {

    private String mail;
    private String password;

    Paypal(String mailP, String pwd) {
        mail = mailP;
        password = pwd;
    }

    @Override
    public void pay(int amountP) {
        System.out.println(amountP + " paid with Paypal.");
    }
}

enum Size {

    S(1), L(2), XL(3);

    private int sizeInt;

    Size(int sizeP) {
        this.sizeInt = sizeP;
    }

    @Override
    public String toString() {
        switch(sizeInt) {
            case 1:
                return "small";
            case 2:
                return "large";
            case 3:
                return "extra large";
            default:
                return "CUSTOM";
        }
    }

}

class Item {

    private Size size;
    private int price;

    Item(Size sizeP, int priceP) {
        size = sizeP;
        price = priceP;
    }

    public int getPrice() {
        return price;
    }
}

public class PayItems {

    private ArrayList<Item> items = new ArrayList<>();

    public void addItem(Item itemP) {
        items.add(itemP);
    }
    public void removeItem(Item itemP) {
        items.remove(itemP);
    }
    public int calculateTotal() {
        int sum = 0;
        for(Item itemP : items) {
            sum += itemP.getPrice();
        }
        return sum;
    }
    public void runPayMethod(PayMethod paymethodP) {
        int amount = calculateTotal();
        paymethodP.pay(amount);
    }

    public static void main(String[] args) throws Exception {
        PayItems pi = new PayItems();
        Item item1 = new Item(Size.S, 20);
        Item item2 = new Item(Size.L, 30);
        Item item3 = new Item(Size.XL, 100);
        pi.addItem(item1);
        pi.addItem(item2);
        pi.runPayMethod(new Paypal("jslfree080@gmail.com", "somePassword"));
        pi.removeItem(item1);
        pi.removeItem(item2);
        pi.addItem(item3);
        pi.runPayMethod(new CreditCard("Jungsoo Lee", 12345678901234L, "080", "12/12"));
    }
}

File: dP/PayItems.kt
interface PayMethod {
    fun pay(amountP: Int)
}

class CreditCard(
    private val name: String,
    private val cardNumber: Long,
    private val cvv: String,
    private val dateOfExpiry: String
) : PayMethod {
    override fun pay(amountP: Int) {
        println("$amountP paid using credit/debit card.")
    }
}

class Paypal(private val mail: String, private val password: String) : PayMethod {
    override fun pay(amountP: Int) {
        println("$amountP paid with Paypal.")
    }
}

enum class Size(private val sizeInt: Int) {
    S(1), L(2), XL(3);

    override fun toString(): String {
        return when (sizeInt) {
            1 -> "small"
            2 -> "large"
            3 -> "extra large"
            else -> "CUSTOM"
        }
    }
}

class Item(private val size: Size, private val price: Int) {
    fun getPrice(): Int {
        return price
    }
}

class PayItems {
    private val items = mutableListOf<Item>()

    fun addItem(itemP: Item) {
        items.add(itemP)
    }
    fun removeItem(itemP: Item) {
        items.remove(itemP)
    }
    fun calculateTotal(): Int {
        var sum = 0
        for (itemP in items) {
            sum += itemP.getPrice()
        }
        return sum
    }
    fun runPayMethod(paymethodP: PayMethod) {
        val amount = calculateTotal()
        paymethodP.pay(amount)
    }
}

fun main() {
    val pi = PayItems()
    val item1 = Item(Size.S, 20)
    val item2 = Item(Size.L, 30)
    val item3 = Item(Size.XL, 100)
    pi.addItem(item1)
    pi.addItem(item2)
    pi.runPayMethod(Paypal("jslfree080@gmail.com", "somePassword"))
    pi.removeItem(item1)
    pi.removeItem(item2)
    pi.addItem(item3)
    pi.runPayMethod(CreditCard("Jungsoo Lee", 12345678901234L, "080", "12/12"))
}